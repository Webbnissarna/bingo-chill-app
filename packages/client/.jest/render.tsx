import React from "react";
import { render as rtlRender } from "@testing-library/react";

export function render(ui: React.ReactElement) {
  return rtlRender(<>{ui}</>);
}
