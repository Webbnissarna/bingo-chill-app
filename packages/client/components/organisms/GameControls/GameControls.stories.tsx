import { useState } from "react";
import GameControls from ".";
import type { GameOptions } from "@/services/GameService/GameService.types";
import type { StoryDefault } from "@ladle/react";

export function Default() {
  const [options, setOptions] = useState<GameOptions>({
    seed: 1337,
    lockout: false,
    tags: [],
    timeLimitMinutes: 0,
  });

  return (
    <div className="bg-polarNight-1 p-5">
      <GameControls
        options={options}
        onChange={setOptions}
        onLoadGameSetup={() => alert("Load Game Setup clicked!")}
        onStartGame={() => alert("Start Game clicked!")}
      />
    </div>
  );
}

export default {
  title: "Organisms / Game Controls",
} satisfies StoryDefault;
