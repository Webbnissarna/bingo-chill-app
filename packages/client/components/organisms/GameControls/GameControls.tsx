import IntegerInputField from "@/components/atoms/IntegerInputField/IntegerInputField";
import Text from "@/components/atoms/Text";
import Button from "@/components/atoms/Button";
import Toggle from "@/components/atoms/Toggle";
import { ServiceRegistryContext } from "@/services/ServiceRegistry/ServiceRegistryContext";
import { useContext } from "react";
import type { SessionOptions } from "@webbnissarna/bingo-chill-common/src/game/types";
import type { DeepPartial } from "@webbnissarna/bingo-chill-common/src/utils/functional";
import { patch } from "@webbnissarna/bingo-chill-common/src/utils/functional";
import { TagsInput } from "@/components/atoms";
import type { Tag } from "@/components/atoms/TagsInput";

interface GameControlsProps {
  options: SessionOptions;
  allTags: Tag[];
  onChange: (newOptions: SessionOptions) => void;
}

export default function GameControls({
  options,
  allTags,
  onChange,
}: GameControlsProps): JSX.Element {
  const serviceRegistry = useContext(ServiceRegistryContext);

  const updateOptions = (update: DeepPartial<SessionOptions>) => {
    onChange(patch(options, update));
  };

  const randomizeSeed = () => {
    const rng = serviceRegistry.get("Randomness");
    const seed = rng.randRangeInt(0, 10000000);
    updateOptions({ seed });
  };

  return (
    <div className="w-full flex flex-col gap-2 items-start">
      <div className="grid grid-cols-[auto_auto] gap-2 w-full">
        <Text>Seed</Text>
        <div className="flex gap-2">
          <IntegerInputField
            value={options.seed}
            onChange={(seed) => updateOptions({ seed })}
          />
          <Button onClick={randomizeSeed}>R</Button>
        </div>

        <Text>Include Tasks</Text>
        <TagsInput
          placeholder="Tags"
          values={options.taskFilters.includedTags}
          options={allTags}
          onChange={(includedTags) =>
            updateOptions({ taskFilters: { includedTags } })
          }
        />
        <Text>Exclude Tasks</Text>
        <TagsInput
          placeholder="Tags"
          values={options.taskFilters.excludedTags}
          options={allTags}
          onChange={(excludedTags) =>
            updateOptions({ taskFilters: { excludedTags } })
          }
        />

        <Text>Lockout?</Text>
        <Toggle
          checked={options.isLockout}
          onChange={(isLockout) => updateOptions({ isLockout })}
        />

        <Text>Time Limit</Text>
        <IntegerInputField
          value={options.timeLimitMinutes}
          onChange={(timeLimitMinutes) => updateOptions({ timeLimitMinutes })}
        />
      </div>
    </div>
  );
}
