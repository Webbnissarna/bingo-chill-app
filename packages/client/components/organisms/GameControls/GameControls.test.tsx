import { render } from "../../../.jest/render";
import GameControls from ".";

describe("GameControls", () => {
  it("renders", () => {
    const noop = () => undefined;
    const { asFragment } = render(
      <GameControls
        options={{ seed: 0, isLockout: false, timeLimitMinutes: 0 }}
        onChange={noop}
        onLoadGameSetup={noop}
        onStartGame={noop}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
