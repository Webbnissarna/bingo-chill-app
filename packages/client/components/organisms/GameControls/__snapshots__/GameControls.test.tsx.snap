// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameControls renders 1`] = `
<DocumentFragment>
  <div
    class="w-full flex flex-col gap-2 items-start"
  >
    <div
      class="grid grid-cols-[auto_auto] gap-2 w-full"
    >
      <span
        class="font-text text-snowStorm-2 text-base"
      >
        Seed
      </span>
      <div
        class="flex gap-2"
      >
        <input
          class="w-full font-text text-base text-snowStorm-2 outline-none border-none shadow-none bg-polarNight-0 rounded px-2 py-1 focus:bg-polarNight-2 transition"
          type="number"
          value="1337"
        />
        <button
          class="font-text font-base text-snowStorm-2 px-3 py-1 rounded-lg transition flex gap-1 items-center justify-center bg-polarNight-3 hover:bg-frost-3 active:bg-frost-2"
        >
          R
        </button>
      </div>
      <span
        class="font-text text-snowStorm-2 text-base"
      >
        Include Tasks
      </span>
      <div
        class="ant-select w-full css-dev-only-do-not-override-18iikkb ant-select-multiple ant-select-allow-clear ant-select-show-arrow ant-select-show-search"
      >
        <div
          class="ant-select-selector"
        >
          <div
            class="ant-select-selection-overflow"
          >
            <div
              class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
              style="opacity: 1;"
            >
              <div
                class="ant-select-selection-search"
                style="width: 0px;"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-label="Search"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  role="combobox"
                  type="search"
                  value=""
                />
                <span
                  aria-hidden="true"
                  class="ant-select-selection-search-mirror"
                >
                   
                </span>
              </div>
            </div>
          </div>
          <span
            class="ant-select-selection-placeholder"
          >
            Tags
          </span>
        </div>
        <span
          aria-hidden="true"
          class="ant-select-arrow"
          style="user-select: none;"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-select-suffix"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
      <span
        class="font-text text-snowStorm-2 text-base"
      >
        Exclude Tasks
      </span>
      <div
        class="ant-select w-full css-dev-only-do-not-override-18iikkb ant-select-multiple ant-select-allow-clear ant-select-show-arrow ant-select-show-search"
      >
        <div
          class="ant-select-selector"
        >
          <div
            class="ant-select-selection-overflow"
          >
            <div
              class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
              style="opacity: 1;"
            >
              <div
                class="ant-select-selection-search"
                style="width: 0px;"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-label="Search"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  role="combobox"
                  type="search"
                  value=""
                />
                <span
                  aria-hidden="true"
                  class="ant-select-selection-search-mirror"
                >
                   
                </span>
              </div>
            </div>
          </div>
          <span
            class="ant-select-selection-placeholder"
          >
            Tags
          </span>
        </div>
        <span
          aria-hidden="true"
          class="ant-select-arrow"
          style="user-select: none;"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-select-suffix"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
      <span
        class="font-text text-snowStorm-2 text-base"
      >
        Lockout?
      </span>
      <input
        class="w-7 h-7 outline-none border-none rounded bg-polarNight-0 text-frost-3 focus:border-none cursor-pointer active:text-frost-2 transition"
        type="checkbox"
      />
      <span
        class="font-text text-snowStorm-2 text-base"
      >
        Time Limit
      </span>
      <input
        class="w-full font-text text-base text-snowStorm-2 outline-none border-none shadow-none bg-polarNight-0 rounded px-2 py-1 focus:bg-polarNight-2 transition"
        type="number"
        value="0"
      />
    </div>
  </div>
</DocumentFragment>
`;
